### ISSUE TITLE
Expression duplication

### ISSUE THREAD

**Issue by dktcoding** [fasseg/exp4j#31]
When using ```exp4j``` to paralellize calculations, a several calls to ```ExpressionBuilder#build()``` are needed (one per thread), but in the current implementation of ```ExpressionBuilder``` the process is repeated over and over again. 
It seems pretty easy to save the parsed tokens internally until the next call to ```ExpressionBuilder#build()``` that way we won't need to go through the whole ```ShuntingYard#convertToRPN()``` again (this shouldn't break anything, at least in the current codebase, if 9a845e1f8a965c03b9bdc02f6d50c9d337aec76c is merged then we'll need to clone variable instances).

This issue was raised in https://groups.google.com/forum/#!topic/exp4j/keeZRiMhFaI

**Comment by fasseg** [comment]
Thanks for opening..
I still have little time on my hands at the moment. But I promise to look at the open issues in exp4j this weekend.

**Comment by fasseg** [comment]
Yeah that should work...lemme implement a PoC and come back to you...

**Comment by fasseg** [comment]
So while playing around with this idea I stumpled upon this:

**But there's a caveat**
Once a `build()` call was successfully completed, all subsequent build calls return the same cached expression. Therefore calling `variable()` `operator()` and `function()` has no effect whatsoever after a successull `build()`. 

This leads to some ugly unexpected behaviour:
```
final ExpressionBuilder b = new ExpressionBuilder("x*y");
Runnable r1 = new Runnable() {
  public void run() {
    b.variable("x").build(); // this should throw an exception, since "y" is not declared
  }
};

Runnable r2 = new Runnable() {
  public void run() {
    b.variables("x","y").build(); 
  }
};
```

Starting r1 might not throw an exception, since r2 might have been come before r1 and correctly declared all variables. This kinda sucks.

Maybe it would be better to let the user decide if he wants to use the cached tokens from the first successful shuntingyard. This would mean exposing a method like `cachedRebuild()`. But I think I like this way better since it does not lead to unexpected behaviour...

**Comment by dktcoding** [comment]
```cachedRebuild()``` seems like a nice idea!

I was thinking along a different line, starting with the *assumption* that there's only one correct way to parse an expression, that being said, once an expression is successfully built there could be no changes on variable names, function declarations or operators (IMHO it makes sense); the current implementation allows redefining a function or operator between builds, and add (even set) inexistent variables names in ```ExpressionBuilder``` (like in [```testExpressionBuilder05()```](https://github.com/fasseg/exp4j/blob/master/src/test/java/net/objecthunter/exp4j/ExpressionBuilderTest.java#L915), noting that this behavior changes with 5457d9139c52acb1011d33ecf025e74d2444e306).

To summarize, my idea is; once an expression is successfully built all the other methods on ```ExpresionBuilder``` should throw an ```IllegalStateException```.

On the other hand, to fix the possible race condition we could add a ```BinarySemaphore``` (or a ```mutex```) and check it on every method call.

BTW since neither ```HashMap``` nor ```HashSet``` are thread safe, and there's no way of knowing if  all the necessary variables and functions are declared before building the expression, it seems that  something like this should be implemented anyway.

**Comment by leogtzr** [comment]
"To summarize, my idea is; once an expression is successfully built all the other methods on ExpresionBuilder should throw an IllegalStateException."

@dktcoding I would suggest to follow the Builder pattern like this:

```java
// Builder Pattern
public class NutritionFacts {
    private final int servingSize;
    private final int servings;
    private final int calories;
    private final int fat;
    private final int sodium;
    private final int carbohydrate;

    public static class Builder {
        // Required parameters
        private final int servingSize;
        private final int servings;

        // Optional parameters - initialized to default values
        private int calories      = 0;
        private int fat           = 0;
        private int carbohydrate  = 0;
        private int sodium        = 0;

        public Builder(int servingSize, int servings) {
            this.servingSize = servingSize;
            this.servings    = servings;
        }

        public Builder calories(int val)
            { calories = val;      return this; }
        public Builder fat(int val)
            { fat = val;           return this; }
        public Builder carbohydrate(int val)
            { carbohydrate = val;  return this; }
        public Builder sodium(int val)
            { sodium = val;        return this; }

        public NutritionFacts build() {
            return new NutritionFacts(this);
        }
    }

    private NutritionFacts(Builder builder) {
        servingSize  = builder.servingSize;
        servings     = builder.servings;
        calories     = builder.calories;
        fat          = builder.fat;
        sodium       = builder.sodium;
        carbohydrate = builder.carbohydrate;
    }
}
```

I mean, 
1) move all the setters from the Expression class to the ExpressionBuilder class
2) Modify public the constructors from the Expression class (public|default --> private)
3) And only return an Expression instance from the ExpressionBuilder.

**Comment by dktcoding** [comment]
@leogtzr the factory pattern is the one currently in use here.

Perhaps the easiest answer is documentation: ```Make sure that all variables, functions and operators are set before the call to ExpressionBuilder#build()``` and just add a ```synchronized``` in ```ExpressionBuilder#build()```

**Comment by leogtzr** [comment]
@dktcoding 
No, it's not ... 
http://www.informit.com/articles/article.aspx?p=1216151&seqNum=2

**Comment by fasseg** [comment]
@dktcoding 
*Throwing `IllegalStateException` after the first succesful `build()` call:*
This will almost certainly introduce a lot of regressions in existing software, since now additional Exceptions are raised.

@leogtzr 
*You suggested:*
> 1. Move all the setters from the Expression class to the ExpressionBuilder class
> 2. Modify public the constructors from the Expression class (public|default --> private)
> 3. And only return an Expression instance from the ExpressionBuilder.

I don't think I understand this correctly, but:
We currently have the `ExpressionBuilder` which adheres to this pattern. 
Making Expression immutable is a no-go, since the user must be able to recalculate via `Expression.evaluate()` with different variable values without retokenizing the expression by calling `ExpressionBuilder.build()`.

**Comment by dktcoding** [comment]
@fasseg You are right, it seems that ```cacheRebuild()``` might be the best approach for existing users. But I still think that there's something fishy about rebuilding expressions.

**Comment by fasseg** [comment]
I too think it's a bit unclean and am still reluctant to implement this since I can't possibly think of all the use cases in multi threaded environments.
 If anyone of you guys has some time on their hands to implement a `cachedRebuild()` for testing purposes, please do so. But this feature will have to be tested thoroughly, preferably by someone who uses exp4j in a multi threaded environment.

**Comment by dktcoding** [comment]
I'll try to work it backwards, by creating a multi threaded test environment first, but it will take me a while since I have a lot of work right now.
